src/hotspot/share/gc/shared/slidingForwarding.cpp

```diff
@@ -130,20 +130,27 @@ uint FallbackTable::home_index(HeapWord* from) {
 
 void FallbackTable::forward_to(HeapWord* from, HeapWord* to) {
   uint idx = home_index(from);
-  HeapWord* found = _table[idx]._from;
-  if (found != nullptr) {
-    if (found != from) {
-      FallbackTableEntry* entry = NEW_C_HEAP_OBJ(FallbackTableEntry, mtGC);
-      entry->_next = _table[idx]._next;
-      entry->_from = _table[idx]._from;
-      entry->_to = _table[idx]._to;
-      _table[idx]._next = entry;
-    } // Else fall-through and update entry below.
-  } else {
-    assert(_table[idx]._next == nullptr, "next-link should be null here");
+  FallbackTableEntry* head = &_table[idx];
+  FallbackTableEntry* entry = head;
+  // Search existing entry in chain starting at idx.
+  while (entry != nullptr) {
+    if (entry->_from == from) {
+      break;
+    }
+    entry = entry->_next;
+  }
+  if (entry == nullptr) {
+    // No entry found, create new one and insert after head.
+    FallbackTableEntry* new_entry = NEW_C_HEAP_OBJ(FallbackTableEntry, mtGC);
+    new_entry->_next = head->_next;
+    new_entry->_from = head->_from;
+    new_entry->_to   = head->_to;
+    head->_next = new_entry;
+    entry = head; // Set from and to fields below.
   }
-  _table[idx]._from = from;
-  _table[idx]._to   = to;
+  // Set from and to in new or found entry.
+  entry->_from = from;
+  entry->_to   = to;
```

Comment:
Why so complicated? Proposal:
```
while (entry != nullptr && entry->_from != from) {
  entry = entry->_next;
}
if (entry == nullptr) {
  FallbackTableEntry* new_entry = NEW_C_HEAP_OBJ(FallbackTableEntry, mtGC);
  new_entry->next = head;
  new_entry->_from = from;
  head = entry = new_entry;
}
entry->_to = to;
```

Reply:
Uhm, so this would not change the actual head

> Why so complicated? Proposal:
> 
> ```
> while (entry != nullptr && entry->_from != from) {
>   entry = entry->_next;
> }
> if (entry == nullptr) {
>   FallbackTableEntry* new_entry = NEW_C_HEAP_OBJ(FallbackTableEntry, mtGC);
>   new_entry->next = head;
>   new_entry->_from = from;
>   head = entry = new_entry;
> }
> entry->_to = to;
> ```

Remember that head points into the array. We cannot actually prepend the new entry, we can only insert it as the first linked entry after head. If I see it correctly, it would not actually change the head-entry (the stuff in the array) except for its _to field. Also, the new_entry would not get linked anywhere. Or what am I missing?

Reply:
Ah, sorry, I just realized you inlined the first row. Okay, never mind then.